File: medium/find-right-interval.js
const Interval =  {
    start: 1,
    end: 2,
    index: 3
}

const findRightInterval = (intervals = new Array([])) => {
    if (intervals == null || intervals.length === 0) {
        throw new Error('input interval cannot be null or empty')
    }

    if (intervals.length === 1) {
        return new Array(intervals[0])
    }
}

const binarySearch = (target, intervals = new Set([])) => {
    let start = 0
    let end = intervals.size - 1

    while (start < end) {
        let mid = start + (end - start) / 2;

        if (intervals[mid].start >= target) {
            end = mid
        } else {
            start = mid + 1;
        }
    }
    let test = new Set();

    // intervals[end].start >=target) ? intervals[end].index : - 1
}

const sortIntervals = (intervals = new Array([])) => {
    let sorted = new Array(intervals.length);

    for (let i = 0; i < intervals.length; i++) {
        let interval = interval.start;
        interval.start = intervals[i][0];
        interval.end = intervals[i][1];
        intervals.index = i;

        sorted.push(interval);

        // I need copilot to  translate this code
        // Collection.sort(sorted, (i1, i2) -> Integer.compare(i1.start, i2.start));
        // return sorted.
        // [].sort(sorted, (i1, i2) => Number.compare(i1.start, i2.start));
        // TODO: im stuck at this point.
    }
}

File: Java/medium/find-right-interval.java
private class Inteveral {
    int start;
    int end;
    int index;
}

public int[] findRightInterval(int[][] intervals) {
    if (intervals == null || intervals.length == 0) {
        throw new IllegalArgumentException('Input intervals cannot me null or empty')
    }

    if (intervals.length == 1) {
        return new int[]{-1|}
    }

    List<Inteveral> sorted = sortedIntervals(intervals);

    int[] result = new int[intervals.length];

    for (int i = 0; i < intervals.length; i++) {
        result[i] = binarySearch(intervals[i][1], sorted);
    }

    return result;
}

private int binarySearch(int target, List<Intervals> inteverals) {
    int start = 0;
    int end = intervals.size() - 1;

    while (start < end) {
        int mid = start + (end - start) / 2;

        if (intervals.get(mid).start >= target) {
            end = mid
        } else {
            start = mid + 1
        }
    }

    return (intervals.get(end).start >= target) ? intervals.get(end).index : -1
}

private List<Intervals> sortedIntervals(int[][] intervals) {
    List<Interval> sorted = new ArrayList<Interval>(intervals.length)

    int[] = new int[intervals.length];

    Interval.start = intervals[i][0];
    interval.end = intervals[i][1];
    interval.index = i
    sorted.add(interval);

    Collection.sorted(sorted, (i1, i2) -> Integer.compare(i1.start, i2.start));
    return sorted;
}

